src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs

```diff
@@ -17,10 +17,19 @@ public class Meter : IDisposable
         private static readonly List<Meter> s_allMeters = new List<Meter>();
         private List<Instrument> _instruments = new List<Instrument>();
         private Dictionary<string, List<Instrument>> _nonObservableInstrumentsCache = new();
+
+        private string? _formattedTags;
```

Comment:
Rather than add this state as internal properties on Meters and Instruments, organizationally I think InstrumentState is the more appropriate abstraction https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Metrics/AggregationManager.cs,282.

InstrumentState represents the MetricsEventSource's private stash of data for the instruments it is actively recording whereas putting this on Instrument/Meter means we are paying several pointers per-instrument as soon as the instrument is defined, regardless whether MetricsEventSource ever listens to it. There isn't currently any MeterState object that corresponds to Meter, but I think it would be fine if you added InstrumentState.MeterTags and the string wound up being duplicated if there are multiple instruments on a Meter that has tags. Its probably almost always null given the rarity of Meter tag usage. For the hashes I suggested below you could emit them directly as ints so I think it would be fast enough to invoke GetHashCode() with no cached data at all for those.

If you are short for time I don't see anything functionally problematic leaving the fields where they are. My primary motivation is preserving the abstraction boundary and then memory overhead is a very distant 2nd.

Reply:
Now I extended the InstrumentState to carry the Unique instrument Id and publishing the Id in all instrument's events.